sklearn/inspect/permutation_importance.py

```diff
@@ -0,0 +1,182 @@
+"""Permutation importance for estimators"""
+from contextlib import contextmanager
+
+import numpy as np
+
+from ..base import is_classifier, clone
+from ..utils import check_random_state
+from ..utils._joblib import Parallel, delayed
+from ..model_selection import check_cv
+from ..metrics import check_scoring
+from ..utils.metaestimators import _safe_split
+
+
+@contextmanager
+def _permute_column(X, column, random_state):
+    """Context manager to permute a column"""
+    original_feature = X[:, column].copy()
+    X[:, column] = random_state.permutation(X[:, column])
+    yield X
+    X[:, column] = original_feature
+
+
+def _fit_and_calcuate_permutation_importance(estimator, X, y, train_indices,
+                                             test_indices, columns, scoring,
+                                             random_state):
+    """Fits and calculates permutation importance
+
+    Fits ``estimator`` on ``X`` and ``y``
+
+    Parameters
+    ----------
+    estimator : object
+        A supervised learning estimator with a `fit` and is compatible with
+        ``scorer``.
+
+    X : array-like, shape = (n_samples, n_features)
+        Training data.
+
+    y : array-like, shape = (n_samples, ...)
+        Target relative to ``X``.
+
+    train_indices : array of int
+        Train indicies.
+
+    test_indices : array of int
+        Test indices.
+
+    columns : list of integers
+        A list of columns to calculate the permutation importance. If `None`,
+        all columns will be used.
+
+    scoring : string, callable or None
+        A string (see model evaluation documentation) or
+        a scorer callable object / function with signature
+        ``scorer(estimator, X, y)``.
+
+    random_state: : RandomState instance
+        Random number generator.
+
+    Returns
+    -------
+    permutation_importance_scores : list
+        Permutation importance scores for each column on the validation set
+        defined by ``test_indices``.
+    """
+    X_train, y_train = _safe_split(estimator, X, y, train_indices)
+    X_test, y_test = _safe_split(estimator, X, y, test_indices, train_indices)
+
+    estimator.fit(X_train, y_train)
+    baseline_score = scoring(estimator, X_test, y_test)
+
+    permutation_importance_scores = []
+    for column in columns:
+        with _permute_column(X_test, column, random_state) as X_perm:
+            feature_score = scoring(estimator, X_perm, y_test)
+            permutation_importance_scores.append(baseline_score -
```

Comment:
What does it mean when this value is negative? Do we need to clip in that case??

Reply:
Negative means that the model performed better with the feature permuted. This could mean that the feature should be dropped.

There is a paragraph about this in https://explained.ai/rf-importance/index.html at Figure 3(a)

Reply:
Interesting. I think both the docstring and the user guide should explain the meaning of negative importance.